modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java

```diff
@@ -1208,12 +1208,33 @@ public IgniteSemaphore semaphore(final String name, @Nullable AtomicConfiguratio
 
                 if (val == null && !create)
                     return null;
-
+           
                 AtomicDataStructureValue retVal = (val == null ? new GridCacheSemaphoreState(cnt,
                     new HashMap<UUID, Integer>(), failoverSafe, ctx.discovery().gridStartTime()) : null);
 
                 GridCacheSemaphoreEx sem0 = new GridCacheSemaphoreImpl(name, key, cache);
 
+                //check Cluster state against semaphore state
+                if(val!=null){ 
+                    ctx.closure().callLocalSafe(
```

Comment:
Actually, there is no need to start a new closure here, since this whole method get() is already executed from within the transaction started in getAtomic()